### ISSUE TITLE
FP redundant_closure on closures that cannot be called directly

### ISSUE THREAD

**Issue by eareimu** [rust-lang/rust-clippy#16641]
### Summary

The error occurred with `LazyCell<impl Fn>`, but I suspect the problem might be with all types where `Deref<Target = impl Fn / FnMut>` (I haven't verified this).

For &Deref<Target = impl Fn/FnMut>, the incorrect advice given by clippy will be different again.

### Lint Name

redundant_closure

### Reproducer

I tried this code:

```rust
let closure = LazyCell::new(|| |x: usize| println!("{x}"));

(0..10).flat_map(|x| (0..10).map(|y| closure(y))).count();
```

I saw this happen:

```
warning: redundant closure
    --> genmeta-common/src/bind.rs:1278:38
     |
1278 |     (0..10).flat_map(|x| (0..10).map(|y| closure(y))).count();
     |                                      ^^^^^^^^^^^^^^ help: replace the closure with the function itself: `&closure`
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure
```

I expected to see this happen:

no warn

### Version

```text
rustc 1.95.0-nightly (b3869b94c 2026-02-23)
binary: rustc
commit-hash: b3869b94cd1ed4bfa2eb28f301535d5e9599c713
commit-date: 2026-02-23
host: x86_64-unknown-linux-gnu
release: 1.95.0-nightly
LLVM version: 22.1.0
```
